File: mobile/scripts/mutation_check.mjs
/**
 * ⛔ Mutation pass over the CLIENT's stated safety properties.
 *
 *     cd mobile
 *     node scripts/mutation_check.mjs
 *     node scripts/mutation_check.mjs emergency-card
 *
 * The backend has `backend/scripts/mutation_check.py`; this is the same idea
 * on this side of the wire, and for the same reason. Four times in one sitting
 * this repository had a test that was green for a reason unrelated to what it
 * claimed, and two of those were here:
 *
 *   - the goals screen's ticks asserted `accessibilityState`, which is on the
 *     element in jsdom whether or not anything reaches the DOM — so a ticked
 *     row and an unticked one announced identically under a green suite
 *   - the source disclosure had the same defect, found by opening a browser
 *
 * A passing suite is evidence about the tests, not about the code. Each entry
 * below breaks a rule CLAUDE.md states in as many words and checks the suite
 * notices. A mutation that SURVIVES is a rule nothing is actually testing.
 *
 * ⛔ IT NEVER EDITS THE WORKING TREE. Everything is copied to a scratch
 * directory first and the suite runs there. The scratch lives under `mobile/`
 * rather than the system temp directory for one reason: jest resolves
 * `node_modules` by walking up from `rootDir`, and `mobile/node_modules` is
 * where it has to land. It is removed afterwards, including on a crash.
 *
 * ⛔ A SURVIVOR IS NOT FIXED BY DELETING THE MUTATION. Fix the test.
 *
 * ⛔ BUT CHECK THE MUTATION ACTUALLY CHANGES BEHAVIOUR FIRST. One that edits
 * the source without changing what it does reports SURVIVED and is
 * indistinguishable here from a rule nothing tests. It has happened on the
 * backend side: `db.query(...).delete()` rewritten as `_unused = db.query(...)`
 * still calls `.delete()` on the same chain, and read as a missing test for a
 * cascade that was working. The anchor count catches a mutation that could not
 * be applied; nothing catches one that applied and meant nothing.
 *
 * Not part of `npm test`: it runs the suite once per mutation.
 */

import { cpSync, rmSync, readFileSync, writeFileSync, existsSync } from "fs";
import { spawnSync } from "child_process";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const MOBILE = dirname(dirname(fileURLToPath(import.meta.url)));
const SCRATCH = join(MOBILE, ".mutation-scratch");

const GOALS = "src/screens/goals/HealthGoalsScreen.tsx";
const CARD = "src/screens/emergency/EmergencyCardScreen.tsx";
const TOKEN = "src/services/tokenStorage.web.ts";
const CARD_STORE = "src/services/emergencyCard.ts";
const NOTIFY = "src/services/notificationService.web.ts";
const TODAY = "src/screens/TodayScreen.tsx";

/**
 * Each entry: the rule, in CLAUDE.md's own words, and the smallest edit that
 * breaks it.
 */
const MUTATIONS = [
  {
    group: "emergency-card",
    // "⛔ An empty field renders as 'Not provided', never as a missing row" —
    // called "the single most important rule on the screen", because a card
    // with no allergies row reads as *no allergies*.
    label: "an empty field renders as a blank instead of \"Not provided\"",
    file: CARD,
    find: "        {provided ? value : NOT_PROVIDED}",
    replace: "        {provided ? value : \"\"}",
    tests: ["__tests__/EmergencyCardScreen.test.tsx"],
  },
  {
    group: "session",
    // "⛔ Do not move the browser's copy to localStorage." CLAUDE.md says
    // both halves are "asserted by tests so that 'fixing the inconsistency'
    // in either direction fails the suite". This checks that claim.
    label: "the session token moves from sessionStorage to localStorage",
    file: TOKEN,
    find: "    return window.sessionStorage ?? null;",
    replace: "    return window.localStorage ?? null;",
    tests: ["__tests__/tokenStorageWeb.test.ts"],
  },
  {
    group: "goals",
    // The same rule on the screen people open every morning. TodayScreen's
    // docstring forbids "missed" in as many words — "MedHelp does not know
    // whether the dose was taken" — and its test asserts the word never
    // appears. But that test's assertions are mostly `queryByText(...)` being
    // null, and a negative assertion passes just as happily when the screen
    // rendered nothing at all. This proves the test fails when the word is
    // really there.
    label: "the today screen calls a passed dose time missed",
    file: TODAY,
    find: '            ? "Earlier today"',
    replace: '            ? "Missed"',
    tests: ["__tests__/TodayScreen.test.tsx"],
  },
  {
    group: "goals",
    // "⛔ This is not an adherence record... No streaks, no percentages, no
    // '3 of 4 done' tiles."
    label: "the goals screen counts how many rows are done today",
    file: GOALS,
    find: "              {goal.activities.map((activity) => (",
    replace:
      "              <Text>{`${goal.activities.filter((a) => a.completedToday).length}" +
      " of ${goal.activities.length} done today`}</Text>}\n" +
      "              {goal.activities.map((activity) => (",
    tests: ["__tests__/GoalScreens.test.tsx"],
  },
  {
    group: "goals",
    // The accessibility rule this session established: state must be in the
    // label, because `accessibilityState` reaches nothing on web.
    label: "the tick's label stops saying whether it is ticked",
    file: GOALS,
    find:
      "                  accessibilityLabel={\n" +
      "                    `${activity.text}, ` +\n" +
      "                    (activity.completedToday ? \"ticked off for today\" : \"not ticked off\")\n" +
      "                  }",
    replace: "                  accessibilityLabel={activity.text}",
    tests: ["__tests__/GoalScreens.test.tsx", "__tests__/accessibleState.test.ts"],
  },
  {
    group: "goals",
    label: "the source disclosure's label stops saying whether it is open",
    file: GOALS,
    find:
      "                      accessibilityLabel={\n" +
      "                        openSources.has(activity.id)\n" +
      "                          ? `Hide where this kind of activity comes from: ${activity.text}`\n" +
      "                          : `Show where this kind of activity comes from: ${activity.text}`\n" +
      "                      }",
    replace:
      "                      accessibilityLabel={`Where this comes from: ${activity.text}`}",
    tests: ["__tests__/GoalScreens.test.tsx", "__tests__/accessibleState.test.ts"],
  },
  {
    group: "goals",
    // "⛔ Folded, not dropped": closed, nothing readable as an endorsement.
    label: "the citation renders even when the disclosure is closed",
    file: GOALS,
    find: "                    {openSources.has(activity.id) && (",
    replace: "                    {true && (",
    tests: ["__tests__/GoalScreens.test.tsx"],
  },
  {
    group: "no-network",
    // "There is no endpoint, no table, and no `fetch` on this path — a
    // test asserts it." The card is the most sensitive thing this app
    // holds and the one thing that has to work with no signal at all.
    label: "the emergency card is posted to a server when it is saved",
    file: CARD_STORE,
    find: "export async function saveCard(card: EmergencyCard): Promise<EmergencyCard> {",
    replace:
      "export async function saveCard(card: EmergencyCard): Promise<EmergencyCard> {" +
      "\n  await fetch('/emergency-card', { method: 'POST', body: JSON.stringify(card) });",
    tests: ["__tests__/emergencyCard.test.ts", "__tests__/EmergencyCardScreen.test.tsx"],
  },
  {
    group: "no-network",
    // "Nothing about a reminder leaves the device. Both services make no
    // network call; a test asserts `fetch` is never called when one fires."
    // A payload naming a person's medication is exactly what would make a
    // push vendor a processor of PHI.
    label: "a fired reminder phones home with the medication name",
    file: NOTIFY,
    find: '    new window.Notification("Time to take your medication", {',
    replace:
      "    void fetch('/telemetry', { method: 'POST', body: reminder.medicationName });" +
      '\n    new window.Notification("Time to take your medication", {',
    tests: ["__tests__/notificationServiceWeb.test.ts"],
  },
];

function copyTree() {
  rmSync(SCRATCH, { recursive: true, force: true });
  // Everything jest needs, and nothing it does not. node_modules is left out
  // deliberately: resolution walks up from the scratch dir into mobile/.
  for (const entry of ["src", "__tests__", "jest.setup.js", "package.json", "tsconfig.json", "babel.config.js", "assets", "app.json"]) {
    const from = join(MOBILE, entry);
    if (existsSync(from)) cpSync(from, join(SCRATCH, entry), { recursive: true });
  }
}

function survives({ file, find, replace, tests }) {
  copyTree();
  const target = join(SCRATCH, file);
  const source = readFileSync(target, "utf8").split("\r\n").join("\n");
  const hits = source.split(find).length - 1;
  if (hits !== 1) {
    // Not a survivor: a mutation that could not be applied proves nothing
    // about the tests, and saying so is the entire point.
    return `the anchor text appears ${hits} times, not once`;
  }
  writeFileSync(target, source.replace(find, replace), "utf8");

  const result = spawnSync(
    process.platform === "win32" ? "npx.cmd" : "npx",
    ["jest", "--rootDir", SCRATCH, "--silent", ...tests],
    { cwd: MOBILE, encoding: "utf8" }
  );
  return result.status === 0 ? "the suite stayed green" : null;
}

const wanted = process.argv[2];

// ⛔ A FILTER THAT MATCHES NOTHING IS AN ERROR, NOT A CLEAN RUN.
//
// This script's entire job is to distrust a green result — its docstring says
// a passing suite is evidence about the tests, not about the code. It failed
// its own standard: `node scripts/mutation_check.mjs --help`, or any typo of a
// group name, matched zero mutations, printed an empty table, reported "Every
// mutation was caught." and exited 0. Nothing had been mutated and nothing
// verified. In CI that is a green tick for work never done, which is the exact
// failure mode the whole script exists to catch.
const GROUPS = [...new Set(MUTATIONS.map((mutation) => mutation.group))];
if (wanted !== undefined && !GROUPS.includes(wanted)) {
  console.error(`unknown group '${wanted}'; choose from ${GROUPS.join(", ")}`);
  process.exit(2);
}

const survivors = [];
let ran = 0;

try {
  console.log("mutation".padEnd(62) + "result".padStart(10));
  console.log("-".repeat(72));
  for (const mutation of MUTATIONS) {
    if (wanted && mutation.group !== wanted) continue;
    ran += 1;
    const reason = survives(mutation);
    console.log(mutation.label.padEnd(62) + (reason ? "SURVIVED" : "caught").padStart(10));
    if (reason) survivors.push([mutation.label, reason]);
  }
} finally {
  rmSync(SCRATCH, { recursive: true, force: true });
}

console.log();
if (survivors.length) {
  console.log("RULES NOTHING IS ACTUALLY TESTING:");
  for (const [label, why] of survivors) console.log(`   - ${label}  (${why})`);
  process.exit(1);
}
// Belt and braces behind the group check above: whatever the reason, a run
// that mutated nothing may never report that everything was caught.
if (ran === 0) {
  console.error("no mutations ran, so nothing was checked.");
  process.exit(2);
}
console.log(`Every mutation was caught (${ran} of ${MUTATIONS.length}).`);

File: backend/scripts/mutation_check.py
"""
Mutation pass over the safety properties of health goals and symptom triage.

    cd <repo root>
    python backend/scripts/mutation_check.py            # everything
    python backend/scripts/mutation_check.py goals      # one group
    python backend/scripts/mutation_check.py triage

For each rule these features claim to enforce, this breaks it and checks the
suite notices. A mutation that SURVIVES is a rule nothing is actually testing.

WHY THIS EXISTS. Four times in one sitting the goals feature had a test that
was green for a reason unrelated to what it claimed:

  - the goals screen's ticks asserted `accessibilityState`, which passes in
    jsdom whether or not anything reaches the DOM
  - the shape bands' attribution rule explained every rejection once the floor
    was set to something unsatisfiable
  - the prompt-ordering test matched a cross-reference rather than a heading
  - and the two complexity tests, found by this script: both used a one-row
    plan, so replacing the discard with `complexity = "moderate"` still failed
    the moderate ROW COUNT. They demonstrated "one row is not three rows"
    while claiming to demonstrate "a missing reading is refused"

A passing suite is evidence about the tests, not about the code. This is the
cheapest way to tell them apart, and it needs no model and no key.

⛔ IT NEVER TOUCHES THE WORKING TREE. Every mutation is applied to a throwaway
COPY of `backend/`, and the suite runs there. An earlier version edited files
in place and restored them in a `finally`, which left a window where an
interrupted run would strand a mutated source file — observed once, on a real
run. That is merely untidy for `goal_structuring.py`. It is unacceptable for
`triage.py`, so the design changed rather than the reassurance.

⛔ THE TRIAGE GROUP READS FENCED MODULES AND CHANGES NONE OF THEM. CLAUDE.md
forbids modifying `triage.py`, `rules_triage.py` and `emergency.py`, and
permits adding tests for them. This adds no test to them and modifies nothing:
it copies them, breaks the copy, and deletes it. The committed files are never
opened for writing — check `git status` after a run and it will say so.

⛔ A SURVIVOR IS NOT FIXED BY DELETING THE MUTATION. Fix the test.

⛔ BUT CHECK THE MUTATION ACTUALLY CHANGES BEHAVIOUR FIRST. A mutation that
edits the source without changing what it does reports SURVIVED and is
indistinguishable, in this output, from a rule nothing tests. It has happened
here: `db.query(...).delete()` was rewritten to `_unused = db.query(...)`,
which still calls `.delete()` on the same chain. That read as a missing test
for a cascade that was working perfectly.

The anchor count catches a mutation that could not be applied. Nothing can
catch one that applied and meant nothing — that is a judgement about the code,
so read the diff a survivor implies before believing it.

⛔ AND CHECK THE GROUP ACTUALLY RUNS THE FILE THAT TESTS THE RULE. Each group
in `SUITES` runs a fixed tuple of test files, so a mutation whose test lives
outside that tuple survives no matter how well tested the rule is — a third
way to read SURVIVED, and the least obvious of the three.

It happened on 2026-09-20. "medication_reminders copies the medication name"
was added to `privacy` and reported SURVIVED, while the test that catches it
sat in `test_reminders_api.py`, which only `integrity` ran. The rule was fully
tested and the tool said nothing was testing it.

So a survivor means one of three things, in the order worth checking:

1. the group does not run the file where the rule is asserted — look at
   `SUITES` first, it is the cheapest to rule out;
2. the mutation applied but changed no behaviour — read the diff it implies;
3. genuinely, nothing tests the rule — and only then, write the test.

Not part of `pytest`: it runs the suite once per mutation, which takes minutes.
"""

import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

for _stream in (sys.stdout, sys.stderr):
    try:
        _stream.reconfigure(encoding="utf-8", errors="replace")
    except (AttributeError, ValueError):  # pragma: no cover - not a tty
        pass

ROOT = Path(__file__).resolve().parent.parent  # backend/

# Paths RELATIVE to backend/, resolved inside the throwaway copy.
CORE = "app/core/goal_structuring.py"
EVID = "app/core/goal_evidence.py"
API = "app/api/goals.py"
TRIAGE = "app/core/triage.py"
RULES = "app/core/rules_triage.py"
IDENTITY = "app/schemas/booking_identity.py"
MAIN = "app/main.py"
SESSION = "app/db/session.py"
APPOINTMENT_MODEL = "app/models/appointment.py"
PROVIDER_LOCATION_MODEL = "app/models/provider_location.py"
REMINDER_MODEL = "app/models/reminder.py"
MEDICATIONS_API = "app/api/medications.py"
GOALS_API = "app/api/goals.py"
REFILL = "app/services/refill_forecast.py"
CONCEPTS = "app/core/symptom_concepts.py"

# Copied per mutation. Nothing here is worth carrying into a scratch tree, and
# a stale __pycache__ would shadow the mutated source.
SKIP = shutil.ignore_patterns("__pycache__", "*.pyc", ".venv", "venv", "*.db",
                              "certs", ".pytest_cache")

# (group, label, file, find, replace) — each a real weakening of a stated rule.
MUTATIONS = [
    (
        "goals",
        "the weekly-shape check is removed entirely",
        CORE,
        "    if days_touched < fewest_days:",
        "    if False:",
    ),
    (
        "goals",
        "the slot ceiling is removed",
        CORE,
        "    if slots > most_slots:",
        "    if False:",
    ),
    (
        "goals",
        "an unrecognised complexity defaults to moderate instead of discarding",
        CORE,
        'return _discard("plan: no recognised reading of the goal\'s complexity")',
        'complexity = "moderate"',
    ),
    (
        "goals",
        "the row count no longer has to match the declared complexity",
        CORE,
        "    if not fewest <= len(raw_activities) <= most:",
        "    if False:",
    ),
    (
        "goals",
        "a row may come back with no detail",
        CORE,
        'return _discard("plan: an activity had no detail")',
        'detail = "x"',
    ),
    (
        "goals",
        "a suggested row is no longer marked generated",
        CORE,
        "                generated=True,",
        "                generated=False,",
    ),
    (
        "goals",
        "a row with no usable day list is kept instead of discarding the plan",
        CORE,
        'return _discard("plan: an activity had no usable list of days")',
        "days = tuple(DAYS)",
    ),
    (
        "goals",
        "a clock time is guessed rather than refused",
        CORE,
        "    return value if _TIME_PATTERN.match(value) else None",
        '    return value if _TIME_PATTERN.match(value) else "08:00"',
    ),
    (
        "goals",
        "the published figure leaves the prompt",
        CORE,
        "- About 150 minutes of moderate, walking-pace activity across the week,",
        "- A sensible amount of walking-pace activity across the week,",
    ),
    (
        "goals",
        "an unknown evidence domain snaps to the nearest source",
        EVID,
        "    return BY_DOMAIN.get(domain.strip().lower())",
        "    return BY_DOMAIN.get(domain.strip().lower()) or _SOURCES[0]",
    ),
    (
        "goals",
        "a citation travels without its caveat",
        API,
        "        caveat=EVIDENCE_CAVEAT,",
        '        caveat="",',
    ),
    (
        "goals",
        "emergency screening no longer runs before the planner",
        API,
        # Unique: the call site, not the import or the module docstring. The
        # bare name appears three times, and the count guard refused it —
        # which is the guard working. The in-place version before it replaced
        # all three without noticing.
        "    guidance = screen_for_emergency(payload.description)",
        "    guidance = None",
    ),
    # -----------------------------------------------------------------------
    # The five properties CLAUDE.md says hold for triage, "each asserted by
    # tests". This asks whether that is true of the tests as written.
    #
    # ⛔ Applied to a COPY. The fenced modules on disk are not written to.
    # -----------------------------------------------------------------------
    (
        "triage",
        "1. the rules default to SELF_CARE instead of URGENT",
        RULES,
        'tier_name="URGENT",\n        reasoning=DEFAULT_REASONING,',
        'tier_name="SELF_CARE",\n        reasoning=DEFAULT_REASONING,',
    ),
    (
        "triage",
        "3. the two layers reconcile with min() instead of max()",
        TRIAGE,
        "    return max(candidates)",
        "    return min(candidates)",
    ),
    (
        "triage",
        "2. an emergency red-flag match returns URGENT instead of EMERGENT",
        RULES,
        '            tier_name="EMERGENT",',
        '            tier_name="URGENT",',
    ),
    (
        "triage",
        "4. the model supplies the reasoning even when its tier lost",
        TRIAGE,
        "    if model_tier is not None and model_tier >= rule_tier and model_reasoning:",
        "    if model_reasoning:",
    ),
    (
        "triage",
        "5. a model outage produces SELF_CARE instead of the rule tier",
        TRIAGE,
        "    model_tier = verdict.tier if verdict else None",
        "    model_tier = verdict.tier if verdict else Tier.SELF_CARE",
    ),
    (
        "triage",
        "3b. the model tier simply replaces the rule tier",
        TRIAGE,
        "    return max(candidates)",
        "    return candidates[-1]",
    ),
    # -------------------------------------------------------------------
    # Data handling. Each is a rule CLAUDE.md states and attaches
    # "a test asserts it" to. Three of them are findings that file lists
    # as CLOSED, so this is the first check that they are closed rather
    # than merely recorded as closed.
    # -------------------------------------------------------------------
    (
        "privacy",
        'a rejected value is echoed back in the validation error',
        MAIN,
        '            "msg": error.get("msg"),',
        '            "msg": error.get("msg"),\n            "input": error.get("input"),',
    ),
    (
        "privacy",
        "BookingIdentity's repr prints its fields",
        IDENTITY,
        '        return "BookingIdentity(<redacted>)"',
        '        return super().__repr__()',
    ),
    (
        "privacy",
        'SQLAlchemy puts bound values back into its exception text',
        SESSION,
        '    hide_parameters=True,',
        '    hide_parameters=False,',
    ),
    (
        "privacy",
        'the appointments table gains a column that could hold an identity',
        APPOINTMENT_MODEL,
        '    __tablename__ = "appointments"',
        '    __tablename__ = "appointments"\n\n    patient_name: Mapped[str | None] = mapped_column(String, nullable=True)',
    ),
    (
        "privacy",
        'provider_locations gains a column saying who looked',
        PROVIDER_LOCATION_MODEL,
        '    __tablename__ = "provider_locations"',
        '    __tablename__ = "provider_locations"\n\n    user_id: Mapped[str | None] = mapped_column(String, nullable=True)',
    ),
    (
        "privacy",
        # ⛔ HEALTH TEXT IN A URL. CLAUDE.md: the user's text reaches this
        # backend by POST, "never as a URL query string, so it stays out of our
        # access logs, proxies, and crash reporters". A query string is the
        # worst place for it — the access log, the reverse proxy, the CDN, the
        # browser history and the next Referer header all write it down, and
        # none of that is under this app's control.
        #
        # `?symptoms=` on a listing endpoint is the obvious way to make a
        # lookup shareable, and would read as entirely reasonable in review.
        "a GET route takes symptom text in the query string",
        "app/api/medications.py",
        "def list_medications(\n",
        "def list_medications(\n    symptoms: str | None = None,\n",
    ),
    (
        "privacy",
        # The third table with a "must not gain a column" rule. The other two
        # were probed here; this one was stated in the model docstring and in
        # CLAUDE.md and checked nowhere, until 2026-09-20.
        #
        # A name column here is the easy, plausible change: it makes the
        # listing query simpler and reads like a denormalisation nobody would
        # question. What it actually does is put the name of a medicine a
        # named person takes into a second table.
        "medication_reminders copies the medication name",
        REMINDER_MODEL,
        '    __tablename__ = "medication_reminders"',
        '    __tablename__ = "medication_reminders"\n\n    medication_name: Mapped[str | None] = mapped_column(String, nullable=True)',
    ),
    # -------------------------------------------------------------------
    # Data that must not outlive what it described, and the lines this
    # app draws around what it is willing to read.
    # -------------------------------------------------------------------
    (
        'integrity',
        'deleting a medication leaves its reminders behind',
        MEDICATIONS_API,
        '    db.query(MedicationReminder).filter(\n        MedicationReminder.medication_id == medication.id\n    ).delete()',
        '    pass',
    ),
    (
        'integrity',
        'deleting a goal leaves its ticks behind',
        GOALS_API,
        # ⛔ Skip the block, do not merely rename it. The first attempt here
        # was `_unused = db.query(...)`, which still runs `.delete()` on the
        # same chain — a semantic no-op that reported SURVIVED and looked
        # exactly like a test gap. See the warning in the module docstring.
        '    activity_ids = [activity.id for activity in goal.activities]\n    if activity_ids:',
        '    activity_ids = [activity.id for activity in goal.activities]\n    if False:',
    ),
    (
        'integrity',
        'the refill forecast starts reading the printed directions',
        REFILL,
        '    quantity_remaining: int | None,',
        '    frequency: str | None = None,\n    quantity_remaining: int | None,',
    ),
    (
        'integrity',
        'a fourth concept combination is added to the emergency combinator',
        CONCEPTS,
        '_COMBINATIONS: tuple[ConceptCombination, ...] = (',
        '_COMBINATIONS: tuple[ConceptCombination, ...] = (\n    ConceptCombination(\n        rule_id="invented_fourth",\n        category="sepsis_meningitis",\n        required=("fever", "rash"),\n        basis="invented by a mutation, which is the point",\n    ),',
    ),
]


SUITES = {
    "goals": ("tests/test_goals.py", "tests/test_goal_evidence.py"),
    "triage": ("tests/test_triage.py", "tests/test_rules_triage.py",
               "tests/test_emergency.py", "tests/test_triage_eval.py"),
    # The data-handling claims. CLAUDE.md attaches "a test asserts it" to
    # each of these; until now nobody had checked whether that was true.
    # ⛔ A MUTATION ONLY MEETS THE FILES ITS GROUP LISTS. A rule tested in a
    # file this tuple omits reports SURVIVED — indistinguishable, in the
    # output, from a rule nothing tests at all. That happened on 2026-09-20:
    # "medication_reminders copies the medication name" survived while the
    # test that catches it sat in `test_reminders_api.py`, which only the
    # `integrity` group ran. Before believing a survivor, check that this
    # tuple actually includes the file where the rule is asserted.
    "privacy": ("tests/test_booking_identity.py",
                "tests/test_security_hardening.py",
                "tests/test_providers_api.py",
                "tests/test_provider_directory.py",
                "tests/test_appointments_api.py",
                "tests/test_intake_api.py",
                # Holds the structural rule that medication_reminders may not
                # copy the medication name — a privacy rule that happens to
                # live in the reminders file.
                "tests/test_reminders_api.py"),
    # Rules about data that must not outlive the thing it described, and
    # about lines this app draws around what it will read.
    "integrity": ("tests/test_medications_api.py",
                  "tests/test_reminders_api.py",
                  "tests/test_goals.py",
                  "tests/test_refill_forecast.py",
                  "tests/test_symptom_concepts.py",
                  "tests/test_protocol_content.py"),
}


def read(path: Path) -> str:
    """Exact text: `newline=""` stops Python translating line endings."""
    with path.open("r", encoding="utf-8", newline="") as handle:
        return handle.read()


def write(path: Path, text: str) -> None:
    with path.open("w", encoding="utf-8", newline="") as handle:
        handle.write(text)


def survives(group: str, relative: str, find: str, replace: str) -> str | None:
    """
    Apply one mutation in a throwaway copy and run that group's suite there.

    Returns None when the suite caught it, or a reason when it did not — which
    includes the anchor text having moved, because a mutation that cannot be
    applied is not a mutation that was caught.
    """
    with tempfile.TemporaryDirectory(prefix="mutation-") as scratch:
        copy = Path(scratch) / "backend"
        shutil.copytree(ROOT, copy, ignore=SKIP)

        target = copy / relative
        # ⛔ Normalised to bare line feeds before matching. These files are
        # CRLF on disk, so a multi-line anchor written with a bare line feed
        # matches nothing — which is how the first triage run reported a
        # mutation it had never actually applied. The copy is deleted either
        # way, so rewriting its line endings costs nothing.
        source = read(target).replace(chr(13) + chr(10), chr(10))
        if source.count(find) != 1:
            # Not a survivor: a mutation that could not be applied proves
            # nothing about the tests, and saying so is the point.
            return f"the anchor text appears {source.count(find)} times, not once"
        write(target, source.replace(find, replace))

        result = subprocess.run(
            [sys.executable, "-m", "pytest", "-q", "--no-header", "-x",
             *SUITES[group]],
            cwd=copy, capture_output=True, text=True,
        )
        return "the suite stayed green" if result.returncode == 0 else None


wanted = sys.argv[1] if len(sys.argv) > 1 else None
if wanted and wanted not in SUITES:
    sys.exit(f"unknown group {wanted!r}; choose from {', '.join(SUITES)}")

survivors = []
print(f"{'mutation':<62}{'result':>10}")
print("-" * 72)
for group, label, relative, find, replace in MUTATIONS:
    if wanted and group != wanted:
        continue
    reason = survives(group, relative, find, replace)
    print(f"{label:<62}{'SURVIVED' if reason else 'caught':>10}")
    if reason:
        survivors.append((label, reason))

print()
if survivors:
    print("RULES NOTHING IS ACTUALLY TESTING:")
    for label, why in survivors:
        print(f"   - {label}  ({why})")
    sys.exit(1)
print("Every mutation was caught.")
